From: Owen Synge <osynge@suse.com>
Date: Wed, 7 Jan 2015 11:36:24 +0100
Subject: [PATCH] radosgw systemd support

Added a first version of radosgw systemd support and associated prestart script.

Signed-off-by: Owen Synge <osynge@suse.com>
---
 src/ceph-rgw-prestart.sh  | 78 +++++++++++++++++++++++++++++++++++++++++++++++
 systemd/ceph-rgw@.service | 15 +++++++++
 2 files changed, 93 insertions(+)
 create mode 100755 src/ceph-rgw-prestart.sh
 create mode 100644 systemd/ceph-rgw@.service

diff --git a/src/ceph-rgw-prestart.sh b/src/ceph-rgw-prestart.sh
new file mode 100755
index 0000000..2ddd8bf
--- /dev/null
+++ b/src/ceph-rgw-prestart.sh
@@ -0,0 +1,78 @@
+#!/bin/sh
+
+eval set -- "$(getopt -o n: --long name:,cluster: -- $@)"
+
+while true ; do
+	case "$1" in
+		-n|--name) name=$2; shift 2 ;;
+		--cluster) cluster=$2; shift 2 ;;
+		--) shift ; break ;;
+	esac
+done
+
+
+# prefix for radosgw instances in ceph.conf
+PREFIX='client.radosgw.'
+
+if [ -z "$name"  ]; then
+    echo "no name paramter"
+    exit 1
+fi
+
+testname=$(ceph-conf --list-sections $PREFIX | grep $name )
+
+if [ -z "$testname"  ]; then
+    echo $name
+    echo "error parsing '$name' : valid types are: $(echo $(ceph-conf --list-sections $PREFIX))"
+    exit 1
+fi
+
+
+RADOSGW=`which radosgw`
+
+if [ -z "$RADOSGW"  ]; then
+    RADOSGW=/usr/bin/radosgw
+fi
+
+if [ ! -x "$RADOSGW" ]; then
+    [ $VERBOSE -eq 1 ] && echo "$RADOSGW could not start, it is not executable."
+    exit 1
+fi
+
+auto_start=`ceph-conf -n $name 'auto start'`
+if [ "$auto_start" = "no" ] || [ "$auto_start" = "false" ] || [ "$auto_start" = "0" ]; then
+  echo "ceph.conf:[$name], says not to start."
+  exit 1
+fi
+
+# is the socket defined?  if it's not, this instance shouldn't run as a daemon.
+rgw_socket=`$RADOSGW -n $name --show-config-value rgw_socket_path`
+if [ -z "$rgw_socket" ]; then
+  echo "socket $rgw_socket could not be found in ceph.conf:[$name], not starting."
+  exit 1
+fi
+
+# mapped to this host?
+host=`ceph-conf -n $name host`
+hostname=`hostname -s`
+if [ "$host" != "$hostname" ]; then
+  echo "hostname $hostname could not be found in ceph.conf:[$name], not starting."
+  exit 1
+fi
+
+user=`ceph-conf -n $name user`
+if [ -n "$user" ]; then
+  if [ "$USER" != "$user" ]; then
+    echo "enviroment \$USER '$USER' does not match '$name' user '$user'"
+    exit 1
+  fi
+fi
+
+
+log_file=`$RADOSGW -n $name --show-config-value log_file`
+if [ -n "$log_file" ]; then
+  if [ ! -e "$log_file" ]; then
+    touch "$log_file"
+  fi
+  chown $USER $log_file
+fi
diff --git a/systemd/ceph-rgw@.service b/systemd/ceph-rgw@.service
new file mode 100644
index 0000000..711ac07
--- /dev/null
+++ b/systemd/ceph-rgw@.service
@@ -0,0 +1,15 @@
+[Unit]
+Description=Ceph rados gateway
+After=network-online.target local-fs.target
+Wants=network-online.target local-fs.target
+PartOf=ceph.target
+
+[Service]
+EnvironmentFile=-/etc/sysconfig/ceph
+Environment=CLUSTER=ceph
+ExecStart=/usr/bin/radosgw -d --name %i
+ExecStartPre=/usr/libexec/ceph/ceph-rgw-prestart.sh --name %i
+User=wwwrun
+
+[Install]
+WantedBy=ceph.target
